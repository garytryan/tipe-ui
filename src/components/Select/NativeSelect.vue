<template>
  <select 
    :disabled="disabled" 
    :data-tipe-ui="$options.name" >
    <option 
      v-if="placeholder"
      :selected="true"
      disabled>{{ placeholder }}</option>
    <option 
      v-for="option in options"
      :value="option.value" 
      :selected="selectedValue.value === option.value"
      :key="option.value"
    >{{ option.label }}</option>
  </select>
</template>


<script>
import vueTypes from 'vue-types'
import SelectOptionShape from '@/types/SelectOptionShape'

export default {
  name: 'TipeNativeSelect',
  props: {
    disabled: vueTypes.bool,
    options: vueTypes.arrayOf(vueTypes.shape(SelectOptionShape)),
    placeholder: vueTypes.string,
    selectedValue: vueTypes.object
  }
}
</script>

<style lang="postcss">
select {
  position: absolute;
  left: -10000px;
  top: auto;
  width: 1px;
  height: 1px;
  overflow: hidden;
}
</style>
