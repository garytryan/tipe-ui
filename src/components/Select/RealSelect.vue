<template functional>
  <select
    :disabled="waiting || disabled"
  >
    <option
      v-if="props.placeholder"
      :selected="true"
      disabled
    >
      {{ placeholder }}
    </option>
    <option
      v-for="option in props.options"
      :value="option.value"
      :selected="props.selectedValue.value === option.value"
      :key="option.value"
    >
      {{ option.label }}
    </option>
  </select>
</template>
