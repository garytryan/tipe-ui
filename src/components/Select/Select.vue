<template>
  <component 
    :data-tipe-ui="$options.name"  
    v-bind="$props" 
    :is="is"/>
</template>

<script>
import vueTypes from 'vue-types'
import SelectOptionShape from '@/types/SelectOptionShape'
import inputProps from '@/types/InputProps'

import TipeSelectUnderline from './SelectUnderline'
import TipeSelectButton from './SelectButton'

export default {
  name: 'TipeSelect',
  components: {
    TipeSelectUnderline,
    TipeSelectButton
  },
  props: {
    type: vueTypes.string,
    options: vueTypes.arrayOf(vueTypes.shape(SelectOptionShape)),
    placeholder: vueTypes.string,
    defaultSelected: vueTypes.shape(SelectOptionShape),
    color: vueTypes.oneOf([
      'default',
      'primary',
      'info',
      'none',
      'success',
      'danger',
      'warning'
    ]),
    ...inputProps
  },
  computed: {
    is() {
      if (this.type === 'button') return 'tipe-select-button'
      return 'tipe-select-underline'
    },
    methods: {
      onChange(val) {
        this.$emit('change', val)
      }
    }
  }
}
</script>
