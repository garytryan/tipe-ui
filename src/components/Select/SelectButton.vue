<template>
  <div 
    :data-tipe-ui="$options.name" 
    class="select-flex"
    @keyup.up="keyup"
    @keyup.down="keydown"
    @keyup.enter="enter"
    v-on="isDisabled ? {} : { click: onClick }"
  >
    <div 
      :class="{[size]: true}" 
      class="select-button">
      <tipe-native-select 
        :options="options" 
        :disabled="isDisabled" 
        :placeholder="placeholder" 
        :selected-value="selectedValue"/>
      <tipe-button 
        :waiting="waiting"
        :color="color" 
        :status="status"
        :size="size" 
        :disabled="isDisabled"
        :dropdown="true"
        icon-after="dropdown">{{ selected }}</tipe-button>
      <tipe-select-dropdown 
        v-if="open" 
        :options="options" 
        :size="size"
        :active-index="activeIndex" 
        @change="onChange"/>
    </div>
  </div>
</template>

<script>
import vueTypes from 'vue-types'
import inputProps from '@/types/InputProps'
import SelectOptionShape from '@/types/SelectOptionShape'
import TipeButton from '../Button'
import TipeSelectDropdown from './SelectDropdown'
import TipeNativeSelect from './NativeSelect'
import {
  selectKeyup,
  selectKeydown,
  selectEnter,
  selectClick,
  selectChange,
  findSelected,
  selectDisabled
} from './select-utils'

export default {
  name: 'TipeSelectButton',
  components: { TipeButton, TipeSelectDropdown, TipeNativeSelect },
  props: {
    placeholder: vueTypes.string,
    options: vueTypes.arrayOf(vueTypes.shape(SelectOptionShape)),
    color: vueTypes
      .oneOf(['default', 'primary', 'info', 'none'])
      .def('default'),
    ...inputProps
  },
  data() {
    return {
      open: false,
      selectedValue: {},
      activeIndex: -1
    }
  },
  computed: {
    isDisabled() {
      return selectDisabled(this)
    },
    selected() {
      return findSelected(this)
    }
  },
  methods: {
    onClick() {
      this.open = selectClick(this)
    },
    onChange(val) {
      this.selectedValue = selectChange(this, val)
    },
    keyup() {
      this.activeIndex = selectKeyup(this)
    },
    keydown() {
      this.activeIndex = selectKeydown(this)
    },
    enter() {
      let { open, activeIndex = this.activeIndex } = selectEnter(this)
      this.open = open
      this.activeIndex = activeIndex
    }
  }
}
</script>

<style scoped lang="postcss">
.select-flex {
  display: flex;
  flex-direction: column;
}

.select-button {
  display: inline-block;
  max-width: 100%;
  position: relative;
  vertical-align: top;

  &.mini {
    height: 1.625rem;
  }
  &.small {
    height: 1.875rem;
  }

  &.medium {
    height: 2.5rem;
  }

  &.large {
    height: 3.125rem;
  }

  &.full {
    height: 3.25rem;
  }
}
</style>
