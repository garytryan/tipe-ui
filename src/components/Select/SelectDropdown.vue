<template>     
  <div 
    class="dropdown" :class="{[size]: true}">
    <div 
      v-for="(option, index) in options"
      :key="option.value"
      :class="{active: index === activeIndex}" 
      tabindex="-1"
      class="dropdown-item"
      @click="$emit('change', option)"><p>{{ option.label }}</p></div>
  </div>
</template>

<script>
import vueTypes from 'vue-types'
import SelectOptionShape from '@/types/SelectOptionShape'

export default {
  name: 'TipeSelectDropdown',
  props: {
    options: vueTypes.arrayOf(vueTypes.shape(SelectOptionShape)),
    size: vueTypes
      .oneOf(['mini', 'small', 'medium', 'large', 'full'])
      .def('full'),
    activeIndex: vueTypes.number
  }
}
</script>

<style scoped lang="postcss">
.dropdown {
  background-color: #fff;
  border-radius: 0 0 5px 5px;
  z-index: 2;
  box-shadow: 5px 20px 40px 0 rgba(0, 0, 0, 0.1);
  padding: 0.6875rem 0;

  &.mini,
  &.small {
    width: 6.5625rem;
  }

  &.medium {
    width: 10rem;
  }

  &.large {
    width: 10.5625rem;
  }

  &.full {
    width: 100%;
  }

  & .dropdown-item {
    height: 2.1875rem;
    display: flex;
    justify-content: center;
    flex-direction: column;

    & p {
      color: #627098;
      font-size: 0.8125rem;
      margin: 0;
      padding-left: 1.0625rem;
    }

    &:hover,
    &.active {
      background-color: #f7f8fb;
      border-left: 3px #1cc5bc solid;

      & p {
        padding-left: calc(1.0625rem - 3px);
      }
    }
  }
}
</style>
