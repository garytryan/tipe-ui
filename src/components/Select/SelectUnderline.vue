<template>
  <div class="fake-select">
    <slot />
  </div>
</template>

<script>
export default {
  name: 'TipeSelectUnderline'
}
</script>

<style scoped lang="postcss">
.fake-select {
  -webkit-appearance: none;
  border: 1px solid transparent;
  border-radius: 0px;
  box-shadow: none;
  display: inline-flex;
  font-size: 1rem;
  height: 3em;
  line-height: 1.5;
  padding-bottom: calc(0.375em - 1px);
  padding-left: 0.25em;
  padding-top: calc(0.375em - 1px);
  position: relative;
  vertical-align: top;
  background-color: #fff;
  border: none;
  border-bottom: 1px solid #b5baca;
  color: #1f346c;
  cursor: pointer;
  font-size: 0.875rem;
  width: 100%;
  outline: 0;
  padding-right: 0.5em;
  -moz-appearance: none;
  appearance: none;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  justify-content: center;

  &.success {
    border-bottom: 1px solid var(--success);
  }
  &.error {
    border-bottom: 1px solid var(--error);
  }
  &.warning {
    border-bottom: 1px solid var(--warning);
  }

  &.small {
    height: 2rem;
  }

  &.large {
    height: 4em;
  }
}
</style>
