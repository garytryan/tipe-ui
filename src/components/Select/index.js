export { default } from './Select.vue'
export TipeSelectButton from './SelectButton.vue'
export TipeSelectUnderline from './Select.vue'
