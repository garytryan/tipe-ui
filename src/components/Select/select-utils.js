export function selectClick(select) {
  return !select.open
}

export function selectChange(select, val) {
  select.$emit('change', val)
  return val
}

export function selectKeyup(select) {
  if (!select.open) {
  } else if (select.activeIndex === 0) return select.options.length - 1
  else return (select.activeIndex -= 1)
}

export function selectKeydown(select) {
  if (!select.open) {
  } else if (select.activeIndex === select.options.length - 1) return 0
  else return (select.activeIndex += 1)
}

export function selectEnter(select) {
  let selectOptions = {}
  if (!select.options) {
  } else if (!select.open) {
    selectOptions.open = true
    return selectOptions
  } else {
    selectChange(select, select.options[select.activeIndex])
    selectOptions.open = false
    selectOptions.activeIndex = -1
    return selectOptions
  }
}

export function findSelected(selected) {
  if (selected.selectedValue.label) return selected.selectedValue.label
  if (selected.placeholder) return selected.placeholder
  if (selected.options) return selected.options[0].label
  return 'Please provide select options'
}

export function selectDisabled(selected) {
  if (selected.waiting || selected.disabled) return true
  return false
}
