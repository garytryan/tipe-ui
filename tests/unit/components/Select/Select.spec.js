import { shallowMount } from '@vue/test-utils'
import { createRenderer } from 'vue-server-renderer'
import Select from '@/components/Select'
import TipeSelectButton from '@/components/Select/SelectButton'
import TipeSelectUnderline from '@/components/Select/SelectUnderline'

const props = {
  options: [
    { label: 'Owner', value: 'owner' },
    { label: 'Admin', value: 'admin' }
  ],
  placeholder: 'testing',
  color: 'primary'
}

describe('Select.vue', () => {
  it('matches previous snapshot', async () => {
    const wrapper = shallowMount(Select)
    await expect(createRenderer().renderToString(wrapper.vm)).toMatchSnapshot()
  })
  it('has correct data-tipe-iu attibute', () => {
    const wrapper = shallowMount(Select)
    expect(wrapper.attributes()['data-tipe-ui']).toBe('TipeSelect')
  })
  it('should have the correct props', () => {
    const wrapper = shallowMount(Select, {
      propsData: {
        ...props,
        type: 'button',
        defaultSelected: { label: 'Owner', value: 'owner' }
      }
    })
    expect(wrapper.props().options).toBe(props.options)
    expect(wrapper.props().type).toBe('button')
    expect(wrapper.props().placeholder).toBe(props.placeholder)
    expect(wrapper.props().color).toBe(props.color)
    expect(wrapper.props().defaultSelected).toEqual({
      label: 'Owner',
      value: 'owner'
    })
  })
  it('should contain the TipeSelectUnderline component', () => {
    const wrapper = shallowMount(Select, {
      propsData: {
        ...props
      }
    })
    expect(wrapper.contains(TipeSelectUnderline)).toBeTruthy()
  })
  it('should contain the TipeSelectButton component', () => {
    const wrapper = shallowMount(Select, {
      propsData: {
        ...props,
        type: 'button'
      }
    })
    expect(wrapper.contains(TipeSelectButton)).toBeTruthy()
  })
})
