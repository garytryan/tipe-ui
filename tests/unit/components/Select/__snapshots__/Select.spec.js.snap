// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select.vue matches previous snapshot 1`] = `Promise {}`;
